Try to make wxWebRequest CI tests more reliable

vadz opened the pull request:
In fact, using Docker might be the best thing to do, actually -- it seems to be available everywhere, so if this still doesn't work, I think I'm going to try doing  this next. These tests are too useful to sacrifice running them automatically entirely...

vadz committed:
Use a different httpbin mirror by default

This one seems slower, but might be more reliable.

vadz committed:
Increase wxWebRequest tests timeout

10s seems to often be not long enough for the Travis/AppVeyor machines.

PBfordev commented on tests/net/webrequest.cpp:
```diff
@@ -107,7 +111,7 @@ class RequestFixture : public wxTimer
     void RunLoopWithTimeout()
     {
-        StartOnce(10000); // Ensure that we exit the loop after 10s.
+        StartOnce(30000); // Ensure that we exit the loop after 10s.
```
```suggestion
        StartOnce(30000); // Ensure that we exit the loop after 30s.
```

vadz replied:
Fixed, thanks.

vadz commented:
Looks like we'll have to run httpbin locally after all. This will mean using HTTP (because I think that httpbin Python module and docker image based on it don't support HTTPS out of the box), but it's still better than nothing.

vadz committed:
Run httpbin locally for the tests in the CI builds

Run httpbin (either directly or inside a container) and set
WX_TEST_WEBREQUEST_URL to point to localhost.

Note that it is important _not_ to have a trailing slash in the root
URL, otherwise WebRequest::Auth::Digest test would fail.

vadz committed:
Use pip for httpbin in Mac Travis CI builds

Docker isn't available under Mac, unfortunately.

Notice that it's still better to use Docker if it is available, rather
than using pip everywhere, as pip has trouble installing httpbin in the
Ubuntu 14.04 build, for example.

vadz committed:
Show WX_TEST_WEBREQUEST_URL being used in the test

Make it clear which URL do we actually use.

vadz commented:
In AppVeyor build, curl works, but our code doesn't, and I have no idea about what could explain it. In principle this could be due to firewall rules, but who would set them up in such a way: to allow connections outside and also allow connections locally using curl, but not to the test itself?

Any help/ideas would be really, really welcome.

MaartenBent commented on build/tools/appveyor-test.bat:
```diff
@@ -1,3 +1,12 @@
+
+pip install httpbin
+start /b python -m httpbin.core
+set WX_TEST_WEBREQUEST_URL="http://127.0.0.1:5000"
```
```suggestion
set WX_TEST_WEBREQUEST_URL=http://127.0.0.1:5000
```
Try this.

Any way to abort the appveyor build early instead of hogging it for 7+ hours?

vadz replied:
I can't believe I didn't notice this :-( Thanks!

> Any way to abort the appveyor build early instead of hogging it for 7+ hours?

I can cancel the builds, but I have no idea why/how is it configured because while I can do it for wx, I can't do it for other GitHub projects. And I don't see any possibility to add someone else to the project in their web UI, so I can't give you the permissions to do it neither, sorry.

MaartenBent replied:
I did not mean to abort it manually. I was wondering about the previous AppVeyor build (https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/37423322)
The test script ends (it shows `!!! Non-GUI test failed`), but then the build does not finish. Maybe you know what is going on?

Maybe you need to stop `python -m httpbin.core`? We'll see what happens to this build.

vadz replied:
Ah, sorry for misunderstanding.

You are probably right and it times out waiting for `python` process to exit. Unfortunately I don't know how to get the PID of the process using `cmd.exe` and using `taskkill /im python` might be too drastic. But short of rewriting the build script in PowerShell (which might not be bad idea...) I don't see anything better, do you?

MaartenBent replied:
You could use `before_test` and `after_test` in `AppVeyor.yml`. Install and start `httpbin`, save the pid as env var, and then stop it after testing. Then you can use PS there, and keep the bat file (for now).

vadz committed:
Fix setting WX_TEST_WEBREQUEST_URL in AppVeyor builds

Don't use quotes around the value, they're not special for cmd.exe
and remain part of the actual variable value.

Co-authored-by: Maarten <MaartenBent@users.noreply.github.com>

MaartenBent commented:
AppVeyor is timing out again.

The teapot test still fails. But httpbin also has a teapot with the following response string:
```wxString teapot = "\n    -=[ teapot ]=-\n\n       _...._\n     .'  _ _ `.\n    | .\"` ^ `\". _,\n    \\_;`\"---\"`|//\n      |       ;/\n      \\_     _/\n        `\"\"\"`\n";```

vadz commented:
> AppVeyor is timing out again.

I've cancelled it for now and will implement your suggestion with `{before ,after}_test` later, thanks.

> The teapot test still fails. But httpbin also has a teapot with the following response string:
> `` wxString teapot = "\n -=[ teapot ]=-\n\n _...._\n .' _ _ `.\n | .\"` ^ `\". _,\n \\_;`\"---\"`|//\n | ;/\n \\_ _/\n `\"\"\"`\n"; ``

Thanks, no idea how I've managed not to notice this one!

vadz committed:
Use httpbin for error status with body test too

Although httpbin doesn't return any body for the other 4xx statuses, it
does make an exception for this one, so we can use it for testing
instead of the even less reliable httpstat.us which regularly returns
503 server error instead of the expected 418.

This probably also makes the workaround for Ubuntu 14.04 unnecessary, so
remove it.

vadz committed:
Run httpbin in appveyor.yml to be able to stop it later

There is no way to stop a background task simply and safely with
cmd.exe, but in appveyor.yml we can use PowerShell, so use its job
support to do it instead.

This is important because AppVeyor waits until all jobs launched by the
script terminate, so without stopping the background job all builds
waited for an hour before timing out.

vadz commented:
Any idea why launching httpbin seems to fail in [MSYS2](https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/37461243/job/splod9wu4a4bggod) (and also in [Cygwin](https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/37461243/job/p8v0msn73ukqdar7), but we don't care about tests failure in this one anyhow) AppVeyor builds?

Also, we should probably switch to Python 3, but how do you tell pip to use [one of the installed](https://www.appveyor.com/docs/windows-images-software/#python) Python 3 versions?

vadz committed:
Try using Python 3 with httpbin

Use 3.5 which is currently available on all AppVeyor Windows images, see
https://www.appveyor.com/docs/windows-images-software/#python

MaartenBent commented:
Msys2 has its own python installation, When only specifying pip, you install it to the Windows version. I think using `python -m pip install httpbin` will fix it.
Specifying which python to run, as you did now, should work as well.

vadz committed:
Suppress pip warning about installing scripts not in PATH

We don't care about it here, as we use full paths anyhow.

vadz commented:
There is a new error in the `PUT` test now :-( This stuff really drives me crazy.

vadz committed:
Show httpbin output if the test failed, not if it succeeded

This output is mostly useful for diagnosing errors, so it's not very
useful to show it only after successful test completion in after_test
command.

vadz committed:
Redirect httpbin error output to the log file too

This can be at least as useful as stdout.

vadz committed:
Disable the tests if launching httpbin failed for some reason

This is not ideal, but will have to do for now.

vadz committed:
Revert "Suppress pip warning about installing scripts not in PATH"

This reverts commit 2eaa7a21b05017969a78198d65f47ad91cd8a356 because the
--no-warn-script-location option only exists in Python 3.9+ apparently
and we don't want to require such a new version just for this.

vadz committed:
Add Python scripts directory to PATH to avoid pip warnings

They're really too annoying.

vadz committed:
Use curl command, not PowerShell alias

Aliasing curl to something completely different (Invoke-WebRequest) is
just evil, yet this is what Powershell 3+ does.

vadz committed:
Fix testing for curl success in PowerShell

Of course doing the natural thing doesn't work in PowerShell.

vadz committed:
Fix return code from testing script in case of error

Using type reset %errorlevel% and the errors were not detected any
longer.

vadz commented:
`PUT` tests keeps sporadically failing for no apparent reason. httpbin log shows 200 return status, yet it somehow fails. And I've just run it 100 times locally but couldn't reproduce the problem. Not sure how to debug it further.

vadz commented:
I'm going to merge this because it still makes things better (currently all CI builds fails because of unreliable external servers), even if it's far from perfect.
